### ISSUE TITLE
Logs clustering 

### ISSUE THREAD

**Issue by FAMILIAR-project** [TuxML/compilation-analysis#1]
We need a procedure to group together failures with the almost same error messages... 

Typically our use case is to investigate whether failures that are supposed to be caused by one option have the same error messages. 

Let's take an example.
We can get all configuration failures in which `AIC7XXX_BUILD_FIRMWARE` is activated
`aic7xx_failures = rawtuxdata.query("AIC7XXX_BUILD_FIRMWARE == 'y' & vmlinux == -1")`
using TUXML-analysis

then we get access to `cid` and retrieve all logs:
```
   logs_ aic7xx_failures = [err_logs_configuration(cid) for cid in aic7xx_failures['cid']]
```
using facilities of `bdd-tuxml-facility`

What remains open is how to cluster logs. 
We can start with the findings of common lines (if any), with pre-defined "error patterns" we know, or more advanced clustering https://scikit-learn.org/0.18/auto_examples/text/document_clustering.html 

Once we have clusters, we have to explain them:
 * is AIC7XXX_BUILD_FIRMWARE really the cause? 
  * in that case, why and to what extend error messages differ?
 * is it due to another (combination of) option? is there a masking effect? 

We can refine the query as well: 
`rawtuxdata.query("AIC7XXX_BUILD_FIRMWARE == 'y' & AIC79XX_BUILD_FIRMWARE == 'y' & vmlinux == -1")` 
to investigate the masking-effect between individual options leading to failures 

Another idea would be to consider all failures
`rawtuxdata.query("vmlinux == -1")` 
and analyze all logs... But too many clusters may popup. We can certainly try, but I suggest to try with individual options first

**Comment by FAMILIAR-project** [comment]
Interesting case... 
There is a configuration with `WANXL_BUILD_FIRMWARE == 'y' & AIC7XXX_BUILD_FIRMWARE == 'y'` (cid=35796)
Which option is faulty? 

According to the decision tree, `AIC7XXX_BUILD_FIRMWARE` but looking at the error log:
`err_logs_configuration(35796)`
> "drivers/gpu/drm/radeon/radeon_object.c: In function 'radeon_bo_create':\ndrivers/gpu/drm/radeon/radeon_object.c:242:2: warning: #warning Please enable CONFIG_MTRR and CONFIG_X86_PAT for better performance thanks to write-combining [-Wcpp]\n #warning Please enable CONFIG_MTRR and CONFIG_X86_PAT for better performance \\\n  ^~~~~~~\n/bin/sh: 1: as68k: not found\nmake[3]: *** [drivers/net/wan/wanxlfw.inc] Error 127\nmake[2]: *** [drivers/net/wan] Error 2\nmake[2]: *** Waiting for unfinished jobs....\nmake[1]: *** [drivers/net] Error 2\nmake[1]: *** Waiting for unfinished jobs....\nmake: *** [drivers] Error 2\n"

It is an error of `WANXL_BUILD_FIRMWARE` (as68k is missing). 

It's different than the basic error/trace of `AIC7XXX_BUILD_FIRMWARE` eg 
err_logs_configuration(70886) 


> 'aicasm_symbol.c:48:19: fatal error: aicdb.h: No such file or directory\n #include "aicdb.h"\n                   ^\ncompilation terminated.\naicasm_gram.tab.c: In function \'yyparse\':\naicasm_gram.tab.c:1628:16: warning: implicit declaration of function \'yylex\' [-Wimplicit-function-declaration]\naicasm_macro_gram.tab.c: In function \'mmparse\':\n/TuxML/linux-4.13.3/drivers/scsi/aic7xxx/aicasm//aicasm_macro_gram.c:64:25: warning: implicit declaration of function \'mmlex\' [-Wimplicit-function-declaration]\n #define yylex           mmlex\n                         ^\naicasm_macro_gram.tab.c:1199:16: note: in expansion of macro \'yylex\'\naicasm_scan.l: In function \'yylex\':\naicasm_scan.l:420:6: warning: implicit declaration of function \'mm_switch_to_buffer\' [-Wimplicit-function-declaration]\n      yy_switch_to_buffer(temp_state);\n      ^~~~~~~~~~~~~~~~~~~\naicasm_scan.l:421:6: warning: implicit declaration of function \'mmparse\' [-Wimplicit-function-declaration]\n      mm_switch_to_buffer(old_state);\n      ^~~~~~~\naicasm_scan.l:424:6: warning: implicit declaration of function \'mm_delete_buffer\' [-Wimplicit-function-declaration]\n      yy_switch_to_buffer(old_state);\n      ^~~~~~~~~~~~~~~~\nmake[4]: *** [aicasm] Error 1\nmake[3]: *** [drivers/scsi/aic7xxx/aicasm/aicasm] Error 2\nmake[2]: *** [drivers/scsi/aic7xxx] Error 2\nmake[1]: *** [drivers/scsi] Error 2\nmake[1]: *** Waiting for unfinished jobs....\nmake: *** [drivers] Error 2\n'



in other words, we cannot only look at the decision tree (or any rule) since we may miss opportunity to find failures caused by other options (AIC7XXX_BUILD_FIRMWARE may mask WANXL_BUILD_FIRMWARE)

**Comment by FAMILIAR-project** [comment]
What I found counter-intuitive and can break some assumptions is this representative example:
`err_logs_configuration(80924)` 
(it's a configuration with "WANXL_BUILD_FIRMWARE == 'y' & AIC7XXX_BUILD_FIRMWARE == 'y'")

the error log is:

> '/bin/sh: 1: as68k: not found\nmake[3]: *** [drivers/net/wan/wanxlfw.inc] Error 127\nmake[3]: *** Waiting for unfinished jobs....\nmake[2]: *** [drivers/net/wan] Error 2\nmake[2]: *** Waiting for unfinished jobs....\naicasm_symbol.c:48:19: fatal error: aicdb.h: No such file or directory\n #include "aicdb.h"\n                   ^\ncompilation terminated.\naicasm_gram.tab.c: In function ‘yyparse’:\naicasm_gram.tab.c:1628:16: warning: implicit declaration of function ‘yylex’ [-Wimplicit-function-declaration]\naicasm_macro_gram.tab.c: In function ‘mmparse’:\n/TuxML/linux-4.13.3/drivers/scsi/aic7xxx/aicasm//aicasm_macro_gram.c:64:25: warning: implicit declaration of function ‘mmlex’ [-Wimplicit-function-declaration]\n #define yylex           mmlex\n                         ^\naicasm_macro_gram.tab.c:1199:16: note: in expansion of macro ‘yylex’\naicasm_scan.l: In function ‘yylex’:\naicasm_scan.l:420:6: warning: implicit declaration of function ‘mm_switch_to_buffer’ [-Wimplicit-function-declaration]\n      yy_switch_to_buffer(temp_state);\n      ^~~~~~~~~~~~~~~~~~~\naicasm_scan.l:421:6: warning: implicit declaration of function ‘mmparse’ [-Wimplicit-function-declaration]\n      mm_switch_to_buffer(old_state);\n      ^~~~~~~\naicasm_scan.l:424:6: warning: implicit declaration of function ‘mm_delete_buffer’ [-Wimplicit-function-declaration]\n      yy_switch_to_buffer(old_state);\n      ^~~~~~~~~~~~~~~~\nmake[4]: *** [aicasm] Error 1\nmake[3]: *** [drivers/scsi/aic7xxx/aicasm/aicasm] Error 2\nmake[2]: *** [drivers/scsi/aic7xxx] Error 2\nmake[2]: *** Waiting for unfinished jobs....\nmake[1]: *** [drivers/scsi] Error 2\nmake[1]: *** Waiting for unfinished jobs....\nmake[1]: *** [drivers/net] Error 2\nmake: *** [drivers] Error 2\n' 

We have the two kinds of errors: as68k missing (WANXL_BUILD_FIRMWARE) and acidb.h missing (purely AIC7XXX_BUILD_FIRMWARE) within the log! 

It means that the compilation process seems to have continued and did not stop at the first error... 
I'm wondering whether it is due to the build process itself or to the environment that tries to install some packages based on compilation errors, re-builds, and so one. 

edit: I've tried for all configurations in which there is `rawtuxdata.query("WANXL_BUILD_FIRMWARE == 'y'")` **all** have somewhere the message `as68k missing` (though the error messages differ) 

Let's see whether we observe similar trends for other options and error messages

**Comment by FAMILIAR-project** [comment]
Another example, maybe more intuitive:

> configuration cid 53234
> /bin/sh: 1: as68k: not found
> make[3]: *** [drivers/net/wan/wanxlfw.inc] Error 127
> make[3]: *** Waiting for unfinished jobs....
> make[2]: *** [drivers/net/wan] Error 2
> make[2]: *** Waiting for unfinished jobs....
> make[1]: *** [drivers/net] Error 2
> make: *** [drivers] Error 2

and that's it, despite the fact there is `rawtuxdata.query("GENERIC_ALLOCATOR == 'n' & DRM_VBOXVIDEO == 'y' & WANXL_BUILD_FIRMWARE == 'y'")`

If we isolate the "bug" `"GENERIC_ALLOCATOR == 'n' & DRM_VBOXVIDEO == 'y'`
(through `rawtuxdata.query("GENERIC_ALLOCATOR == 'n' & DRM_VBOXVIDEO == 'y' & WANXL_BUILD_FIRMWARE != 'y' & AIC7XXX_BUILD_FIRMWARE != 'y' & AIC79XX_BUILD_FIRMWARE != 'y'")['cid'].values`) we can get a much clearer error message directly related to GENERIC_ALLOCATOR and DRM_VBOXVIDEO: 

> configuration cid 99909
> drivers/staging/vboxvideo/vbox_hgsmi.o: In function `hgsmi_buffer_alloc':
> vbox_hgsmi.c:(.text+0x24): undefined reference to `gen_pool_dma_alloc'
> drivers/staging/vboxvideo/vbox_hgsmi.o: In function `hgsmi_buffer_free':
> vbox_hgsmi.c:(.text+0x105): undefined reference to `gen_pool_free'
> drivers/staging/vboxvideo/vbox_hgsmi.o: In function `hgsmi_buffer_submit':
> vbox_hgsmi.c:(.text+0x11e): undefined reference to `gen_pool_virt_to_phys'
> drivers/staging/vboxvideo/vbox_main.o: In function `vbox_hw_fini':
> vbox_main.c:(.text+0x14a): undefined reference to `gen_pool_destroy'
> drivers/staging/vboxvideo/vbox_main.o: In function `vbox_driver_load':
> vbox_main.c:(.text+0x561): undefined reference to `gen_pool_create'
> vbox_main.c:(.text+0x58e): undefined reference to `gen_pool_add_virt'
> vbox_main.c:(.text+0x59e): undefined reference to `gen_pool_destroy'
> make: *** [vmlinux] Error 1

basically here `WANXL_BUILD_FIRMWARE` masks the error of `GENERIC_ALLOCATOR + DRM_VBOXVIDEO`

**Comment by FAMILIAR-project** [comment]
I've made a proper build of cid=80924 

By proper I mean:
 * same configuration
 * but _not_ using the try-and-fix of TuxML that may re-compile once a learning error has been observed
 * instead, using the classical build process of Linux 
 
full log here (output + err messages):
https://gist.github.com/FAMILIAR-project/58aeb0aaf1d476148fd4430f0b63136e 

```
/bin/sh: 1: as68k: not found
drivers/net/wan/Makefile:58: recipe for target 'drivers/net/wan/wanxlfw.inc' failed
make[3]: *** [drivers/net/wan/wanxlfw.inc] Error 127
scripts/Makefile.build:561: recipe for target 'drivers/net/wan' failed
make[2]: *** [drivers/net/wan] Error 2
scripts/Makefile.build:561: recipe for target 'drivers/net' failed
make[1]: *** [drivers/net] Error 2
Makefile:1019: recipe for target 'drivers' failed
make: *** [drivers] Error 2
```

it does validate our hypothesis: 
 * Linux build stops at the first error 
 * it may mask/hide the error of aic7xxx option 
 * it's a nice feature of TuxML since we gather much more information 

(I've made similar experiments with other configurations) 
related to issue https://github.com/TuxML/ProjetIrma/issues/137

**Comment by arnobl** [comment]
So it means that we cannot really observe the masking effect since builds stop at the first error?
